### DISCUSSION TITLE
Make an option to provide the volume plugin type/name for all the volumes (to aid backups)

### DISCUSSION THREAD

**Discussion by Bubbelb** [nextcloud/all-in-one#5617] [Ideas] [answered]
Well, see it as part of a bigger idea: I'm using BTRFS as storage backend for containers. BTRFS in this way is only of benefit for the images and the dynamic data in the container, built of that image.

But I want more: I want to make snapshots, or clones of all the volumes, so that I can make the following process work:

1. Stop all the (Nextcloud all-in-one) containers
2. Snapshot/Clone alle the volumes that are to be backed-up
3. Start all the (Nextcloud all-in-one) containers
4. Run Borgbackup to backup all the snapshots/clones
5. Remove the snapshots/clones
6. rinse-repeat :)

To this end I'm planning to try out the 'buttervolume' Docker volume plugin. See [here](https://github.com/ccomb/buttervolume)

The big benefit of this is the absolute minimal downtime for a backup, even when there are big volumes of data being ingested, which obviously require more backup time.

A secondary benefit can be that the snapshots can be used for instant recovery, in stead of the borgbackups. I think this would increase restore times. Though I'd still recommend that the borgbackups remain, so one can also recover from a disk, or filesystem failure, using the borgbackups.

What do you all think of it?

**Comment by szaimen** [comment, marked as answer]
Hi, this is already possible via https://github.com/nextcloud/all-in-one/tree/main/manual-install. See https://github.com/nextcloud/all-in-one/blob/main/manual-install/latest.yml

**Reply by Bubbelb** [reply]
Yes, you are right, but that route disables a lot of functionalities I like about AIO, especially all the functions of the management container.

And I'd like to try and make the mentioned backup procedure automated in a community-container. I know this wold require some work in the management container, since it overrides the standard backup.

But that is all future planning for now. So i'll create a test bed with a manual install, and try to create the backup in the way that is mentioned here. I'll get back when I get it working.

**Reply by Bubbelb** [reply]
Well, this is a much harder nut to crack, than I thought. I investigated some solutions, but none of them provide an easy way of installing, maintaining and running stable enough to be put into the backup/restore process.

The first two solutions (buttervolume and docker-btrfs-plugin) both could provide a direct answer to my question, but are unworkable, due to other reasons. The other ones are a slight deviation, with the same aim to provide a quick snapshot of docker volumes for backup.

Here an overview of what I tried, and why it fails.

### buttervolume
https://github.com/ccomb/buttervolume

Is a volume management plugin that has a lot of features. One of them is creating a snapshot volume of an existing BTRFS volume.

But does not start on a RPi. So no-go.

### docker-btrfs-plugin
https://github.com/ccomb/buttervolume

Is a simpler volume management plugin for BTRFS than buttervolume, but supports snapshotting of a volume.
The software also keeps a hard link between original volume and it's snapshots. And the original cannot be removed, when there are snapshots. This makes replacing the original with a backup more complex.

It can only be manual installed and activity in the GIT repo is somewhat stalled. So a no-go here

### using btrfs tools in container

BTRFS is not restricted inside containers, in respect to handling snapshots. So we can make two docker volumes, which reside on the same filesystem. Then we can use btrfs tools to create a snapshot of a subvolume in one docker volume and put it in another docker volume. This works well.

But, there is no source btrfs subvolume available to snapshot in the first place. Docker volumes are created in a btrfs subvolume and the underlying _data folder is created and that is the root of the docker volume.

In Docker Compose, this can be worked around, by using ''subpath'' in a volume. See [here](https://docs.docker.com/reference/compose-file/services/#long-syntax-5). The volume must be created, before the actual docker container is started. And in that volume a btrfs subvolume must be created with the same name as stated in the subpath. To summarise:
1. create docker volume nextcloud_data
2. create temporary container with nextcloud_data volume
  1. do ''btrfs subvol create subpath'' in the root of the nextcloud_data volume
3. create the actual container (via compose?) with subpath entry.

Now, we can mount nextcloud_data in another container, without the subpath directive. This then shows the subpath btrfs subvolume. And now it is possible to create a btrfs snapshot of it. It even can stay in the same docker volume, since it is not visible for the normal containers, because of their subpath configuration.

Though this is still the most viable solution, in my view, it requires quite an overhaul in how docker volumes are created and administered in AIO. Think of migration paths, if this would be the desired solution. And I don't see any subpath options in Docker CLI, so I don't know if it is even possible to get it to work.

The upside is that you can swap nearly instantly between backups, because this only involves (re)moving the current btrfs subvolume and replace it with a snapshot of the desired snapshot.

So the amount of rework that is needed, seems a bridge too far for me. But @szaimen, if you're are interested, I'm glad to help out, where I can.

### Using reflink cloning 
The idea is to use rclone, or rsync to update a second (clone) volume with reflinks.

Reflinks are somewhat similar to hard links. But when you modify a file with two hardlinks, the file gets modified on both places. With reflinks only one gets updated. Effectively creating a snapshot of that file.

The benefit of this approach is that it dos not change anything in the original docker volumes, but only requires separate docker volume for each docker volume to back up, to store the snapshot in.

The drawback of this approach is that... it doesn't work. Both the rsync and rclone projects are looking into reflink support (rclone has it already, but only for Mac), but btrfs support is not functional for both.

Using the standard ''cp'' utility with --reflink=always could be of some help, but it would require an extra delete-pass of some sort, because ''cp'' does not provide in removal of target files that are not present anymore at the source.

Another issue is time. Btrfs snapshots are created in mere seconds. Irrespective of the size, or number of files. cp, rclone and rsync need to traverse through all the folders to check for changes. This takes minutes for every pass.

## Summary

1. The docker volume plugins for btrfs are not (yet) up to standard to deliver a robust base for btrfs snapshots of docker volumes.
2. Using subpaths requires a decent amount of rework and probably conversion built into the aio code.
3. Using reflinks is slow and the tooling is not available yet.

Hopes this helps a bit in knowing wat is an is not possible.

Greetings, Bubbel

**Reply by chri2** [reply]
I ended up here, because I need to extend the storage for my Nextcloud-AIO which is already on a disk using btrfs.

I looked through the documentation of docker about storage drivers and volumes, ended up finding buttervolume and looked for references of people having docker volumes as btrfs subvolumes.

Thanks for this really helpful and detailed information about checking the options.

For the same reasons mentioned I'd love to stick with AIO and not change to a manual install.

My goal is to be able to snapshot the docker volumes used by Nextcloud AIO individually for all the reasons mentioned already.

Did you (or anybody coming along here) test to preconfigure the btrfs filesystem with subvolumes for the Nextcloud AIO docker volumes (in my setup `/var/lib/docker/volumes/nextcloud_*` and `nc_app_*`)?

I'm suspecting that this wouldn't make a difference to docker and the default volume driver nor to Nextcloud AIO. My hope would be that it'd allow to _stop containers through the AIO interface_ and then take snapshots for easy rollbacks.
